eclipse-theia/theia
# electron: use inversify in main process

Closes #3364
Closes #7406
Closes #7964

Combined work with @kittaakos.

#### What it does


This PR contributes a JSON-RPC communication scheme between the `electron-main` process and the `renderer` processes. It is reusing the same architecture as the websocket-based connection between frontend/backend, but adapted to work over electron's ipc apis. Currently there isn't scoped connections, meaning that each renderer process will interact with the same singleton in the electron-main process, but for the current use-cases this is sufficient.

#### How to test


To verify:
 - `yarn --cwd ./examples/electron start` -> Should open previous workspace, if any. Otherwise, opens a no-workspace app.
 - `yarn --cwd ./examples/electron start ~/Desktop/` -> Opens your app with a workspace to the `Desktop`.
 - `yarn --cwd ./examples/electron start ../..` -> Opens the app to with Theia as the workspace.
 - `yarn --cwd ./examples/electron start path/to/missing` -> Opens the app without a workspace. See log: `Could not resolve the workspace path.`
 - `mkdir ~/Desktop/path\ with\ space && yarn --cwd ./examples/electron start ~/Desktop/path\ with\ space` -> Open the app with the desired workspace. The workspace contains spaces in the FS path.

----

 - Check updater sample
 - Check `File` > `Settings` > `Check for Update...` -> You will see, there are no updates.
 - Execute the _Electron Updater Sample: Mock - Available_ command and wait a bit.
 - You should receive a notification, about new updates. Click on yes, the electron-main process should print something to the backend console.
 - Refresh the browser window, you should see `Disposed a sample updater client.` AND `Registered a new sample updater client.`
 - Open a second workspace, you should see `Registered a new sample updater client.`. Close the new window, you should see `Disposed a sample updater client.`

#### Review checklist


#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
